test/jdk/java/net/httpclient/ManyRequests.java

```diff
@@ -139,60 +155,118 @@ protected void close(HttpExchange t, InputStream is) throws IOException {
         int port = server.getAddress().getPort();
-        URI baseURI = new URI("https://localhost:" + port + "/foo/x");
+
+        URI baseURI = URIBuilder.newBuilder()
+                .scheme("https")
+                .host(InetAddress.getLoopbackAddress().getHostName())
+                .port(port)
+                .path("/foo/x").build();
         server.createContext("/foo", new TestEchoHandler());
         server.start();
 
-        RequestLimiter limiter = new RequestLimiter(40);
-        Random rand = new Random();
-        CompletableFuture<?>[] results = new CompletableFuture<?>[REQUESTS];
-        HashMap<HttpRequest,byte[]> bodies = new HashMap<>();
-
-        for (int i=0; i<REQUESTS; i++) {
-            byte[] buf = new byte[(i+1)*CHUNK_SIZE+i+1];  // different size bodies
-            rand.nextBytes(buf);
-            URI uri = new URI(baseURI.toString() + String.valueOf(i+1));
-            HttpRequest r = HttpRequest.newBuilder(uri)
-                                       .header("XFixed", "true")
-                                       .POST(BodyPublishers.ofByteArray(buf))
-                                       .build();
-            bodies.put(r, buf);
-
-            results[i] =
-                limiter.whenOkToSend()
-                       .thenCompose((v) -> {
-                           System.out.println("Client: sendAsync: " + r.uri());
-                           return client.sendAsync(r, BodyHandlers.ofByteArray());
-                       })
-                       .thenCompose((resp) -> {
-                           limiter.requestComplete();
-                           if (resp.statusCode() != 200) {
-                               String s = "Expected 200, got: " + resp.statusCode();
-                               System.out.println(s + " from "
-                                                  + resp.request().uri().getPath());
-                               return completedWithIOException(s);
-                           } else {
-                               counter++;
-                               System.out.println("Result (" + counter + ") from "
-                                                   + resp.request().uri().getPath());
-                           }
-                           return CompletableFuture.completedStage(resp.body())
-                                      .thenApply((b) -> new Pair<>(resp, b));
-                       })
-                      .thenAccept((pair) -> {
-                          HttpRequest request = pair.t.request();
-                          byte[] requestBody = bodies.get(request);
-                          check(Arrays.equals(requestBody, pair.u),
-                                "bodies not equal:[" + bytesToHexString(requestBody)
-                                + "] [" + bytesToHexString(pair.u) + "]");
-
-                      });
-        }
+        boolean done = false;
+        LOOP: do {
```

Comment:
Could we add a comment here indicating the purpose of the loop in terms of catching ConnectException (only) and what happens in terms of the reduction of the request concurrency?